k2/python/k2/rnnt_loss.py

```diff
@@ -0,0 +1,995 @@
+        we still need this for position S is that we may still be emitting
+        the termination/next-frame symbol at this point.
+      am:
+        Acoustic-model part of un-normalized logprobs of symbols, to be added
+        to language-model part before normalizing.  Of shape:
+           [B][T][C]
+        where B is the batch size, T is the maximum sequence length of
+        the acoustic sequences (in frames); and C is size of the symbol
+        vocabulary, including the termination/next-frame symbol.  It reflects
+        the "acoustic" part of the probability of any given symbol appearing
+        next on this frame.
+      symbols:
+        A LongTensor of shape [B][S], containing the symbols at each position
+        of the sequence, possibly including EOS
+      termination_symbol:
+        The identity of the termination symbol, must be in {0..C-1}
+      boundary:
+        a LongTensor of shape [B, 4] with elements interpreted as
+        [begin_symbol, begin_frame, end_symbol, end_frame] that is treated as
+        [0, 0, S, T]
+        if boundary is not supplied.
+        Most likely you will want begin_symbol and begin_frame to be zero.
+    Returns:
+        (px, py) (the names are quite arbitrary).
+           px: logprobs, of shape [B][S][T+1]
+           py: logprobs, of shape [B][S+1][T]
+        in the recursion:
+          p[b,0,0] = 0.0
+          p[b,s,t] = log_add(p[b,s-1,t] + px[b,s-1,t],
+                             p[b,s,t-1] + py[b,s,t-1])
+          .. where p[b][s][t] is the "joint score" of the pair of subsequences
+          of length s and t respectively.  px[b][s][t] represents the
+          probability of extending the subsequences of length (s,t) by one in
+          the s direction, given the particular symbol, and py[b][s][t]
+          represents the probability of extending the subsequences of length
+          (s,t) by one in the t direction,
+          i.e. of emitting the termination/next-frame symbol.
+
+          px[:,:,T] equals -infinity, meaning on the "one-past-the-last" frame
+          we cannot emit any symbols.  This is simply a way of incorporating
+          the probability of the termination symbol on the last frame.
+    """
+    assert (
+        lm.ndim == 3
+        and am.ndim == 3
+        and lm.shape[0] == am.shape[0]
+        and lm.shape[2] == am.shape[2]
+    )
+    (B, T, C) = am.shape
+    S = lm.shape[1] - 1
+    assert symbols.shape == (B, S)
+
+    # subtracting am_max and lm_max is to ensure the probs are in a good range
+    # to do exp() without causing underflow or overflow.
+    am_max, _ = torch.max(am, dim=2, keepdim=True)  # am_max: [B][T][1]
+    lm_max, _ = torch.max(lm, dim=2, keepdim=True)  # lm_max: [B][S+1][1]
+    am_probs = (am - am_max).exp()
+    lm_probs = (lm - lm_max).exp()
+    # normalizers: [B][S+1][T]
+    normalizers = (
+        torch.matmul(lm_probs, am_probs.transpose(1, 2)) + 1.0e-37
```

Comment:
What is the magic number `1.0e-37`?

Reply:
It's close to the lowest representable single precision float.  I was concerned making this too high, like 1e-20, might be the cause of instability.

Reply:
There is 
```
torch.finfo(torch.float32).tiny
```
which is `1.17549e-38`, Is this the value you want?

Reply:
Yes I suppose so.  Could even use the dtype of lm_probs instead of torch.float32.